File: frontend/src/services/notificationService.js
import api from './axiosClient';

const NOTIFICATION_API = '/notifications';

export const notificationService = {
    // Lấy tất cả notification của user
    getUserNotifications: async (page = 0, size = 20, storeId = null) => {
        try {
            const params = new URLSearchParams();
            params.append('page', page);
            params.append('size', size);
            if (storeId) {
                params.append('storeId', storeId);
            }
            
            const response = await api.get(`${NOTIFICATION_API}?${params.toString()}`);
            return response.data;
        } catch (error) {
            console.error('Error fetching notifications:', error);
            throw error;
        }
    },

    // Lấy số notification chưa đọc của user
    getUnreadCount: async (storeId = null) => {
        try {
            const params = storeId ? `?storeId=${storeId}` : '';
            const response = await api.get(`${NOTIFICATION_API}/unread-count${params}`);
            return response.data.unreadCount;
        } catch (error) {
            console.error('Error fetching unread count:', error);
            return 0;
        }
    },

    // Lấy tất cả notification của store (cho Staff xem)
    getStoreNotifications: async (page = 0, size = 20, storeId) => {
        try {
            const params = new URLSearchParams();
            params.append('page', page);
            params.append('size', size);
            params.append('storeId', storeId);
            
            const response = await api.get(`${NOTIFICATION_API}/store?${params.toString()}`);
            return response.data;
        } catch (error) {
            console.error('Error fetching store notifications:', error);
            throw error;
        }
    },

    // Lấy số notification chưa đọc của store (cho Staff xem)
    getStoreUnreadCount: async (storeId) => {
        try {
            const response = await api.get(`${NOTIFICATION_API}/store/unread-count?storeId=${storeId}`);
            return response.data.unreadCount;
        } catch (error) {
            console.error('Error fetching store unread count:', error);
            return 0;
        }
    },

    // Lấy tất cả notification của tất cả store (cho Admin xem)
    getAllStoreNotifications: async (page = 0, size = 20) => {
        try {
            const params = new URLSearchParams();
            params.append('page', page);
            params.append('size', size);
            
            const response = await api.get(`${NOTIFICATION_API}/all-stores?${params.toString()}`);
            return response.data;
        } catch (error) {
            console.error('Error fetching all store notifications:', error);
            throw error;
        }
    },

    // Lấy số notification chưa đọc của tất cả store (cho Admin xem)
    getAllStoreUnreadCount: async () => {
        try {
            const response = await api.get(`${NOTIFICATION_API}/all-stores/unread-count`);
            return response.data.unreadCount;
        } catch (error) {
            console.error('Error fetching all store unread count:', error);
            return 0;
        }
    },

    // Đánh dấu notification đã đọc
    markAsRead: async (notificationId) => {
        try {
            const response = await api.put(`${NOTIFICATION_API}/${notificationId}/read`);
            return response.data;
        } catch (error) {
            console.error('Error marking notification as read:', error);
            throw error;
        }
    },

    // Đánh dấu tất cả notification đã đọc
    markAllAsRead: async (storeId = null) => {
        try {
            const params = storeId ? `?storeId=${storeId}` : '';
            const response = await api.put(`${NOTIFICATION_API}/mark-all-read${params}`);
            return response.data;
        } catch (error) {
            console.error('Error marking all notifications as read:', error);
            throw error;
        }
    },

    // Đánh dấu tất cả notification của tất cả store đã đọc (cho Admin)
    markAllNotificationsAsRead: async () => {
        try {
            const response = await api.put(`${NOTIFICATION_API}/mark-all-notifications-read`);
            return response.data;
        } catch (error) {
            console.error('Error marking all notifications as read:', error);
            throw error;
        }
    },

    // Xóa notification
    deleteNotification: async (notificationId) => {
        try {
            const response = await api.delete(`${NOTIFICATION_API}/${notificationId}`);
            return response.data;
        } catch (error) {
            console.error('Error deleting notification:', error);
            throw error;
        }
    },

    // Xóa tất cả notification
    deleteAllNotifications: async (storeId = null) => {
        try {
            const params = storeId ? `?storeId=${storeId}` : '';
            const response = await api.delete(`${NOTIFICATION_API}/delete-all${params}`);
            return response.data;
        } catch (error) {
            console.error('Error deleting all notifications:', error);
            throw error;
        }
    },

    // Tạo notification mới
    createNotification: async (notificationData) => {
        try {
            const response = await api.post(NOTIFICATION_API, notificationData);
            return response.data;
        } catch (error) {
            console.error('Error creating notification:', error);
            throw error;
        }
    },

    // Tạo notification cho tất cả user trong store
    createNotificationForStore: async (storeId, notificationData) => {
        try {
            const response = await api.post(`${NOTIFICATION_API}/store/${storeId}`, notificationData);
            return response.data;
        } catch (error) {
            console.error('Error creating notification for store:', error);
            throw error;
        }
    },

    // Tạo notification cho import transaction
    createImportTransactionNotification: async (action, transactionName, storeId, userId, newStatus = null) => {
        try {
            const response = await api.post(`${NOTIFICATION_API}/import-transaction`, {
                action,
                transactionName,
                storeId,
                userId,
                newStatus
            });
            return response.data;
        } catch (error) {
            console.error('Error creating import transaction notification:', error);
            throw error;
        }
    },

    // Tạo notification cho sale transaction
    createSaleTransactionNotification: async (action, transactionName, storeId, userId, newStatus = null) => {
        try {
            const response = await api.post(`${NOTIFICATION_API}/sale-transaction`, {
                action,
                transactionName,
                storeId,
                userId,
                newStatus
            });
            return response.data;
        } catch (error) {
            console.error('Error creating sale transaction notification:', error);
            throw error;
        }
    },

    // Tạo notification cho product
    createProductNotification: async (action, productName, storeId, userId) => {
        try {
            const response = await api.post(`${NOTIFICATION_API}/product`, {
                action,
                productName,
                storeId,
                userId
            });
            return response.data;
        } catch (error) {
            console.error('Error creating product notification:', error);
            throw error;
        }
    },

    // Tạo notification cho customer
    createCustomerNotification: async (action, customerName, storeId, userId) => {
        try {
            const response = await api.post(`${NOTIFICATION_API}/customer`, {
                action,
                customerName,
                storeId,
                userId
            });
            return response.data;
        } catch (error) {
            console.error('Error creating customer notification:', error);
            throw error;
        }
    },

    // Tạo notification cho stocktake
    createStocktakeNotification: async (action, stocktakeName, storeId, userId, newStatus = null) => {
        try {
            const response = await api.post(`${NOTIFICATION_API}/stocktake`, {
                action,
                stocktakeName,
                storeId,
                userId,
                newStatus
            });
            return response.data;
        } catch (error) {
            console.error('Error creating stocktake notification:', error);
            throw error;
        }
    },

    // Tạo notification cho category
    createCategoryNotification: async (action, categoryName, storeId, userId) => {
        try {
            const response = await api.post(`${NOTIFICATION_API}/category`, {
                action,
                categoryName,
                storeId,
                userId
            });
            return response.data;
        } catch (error) {
            console.error('Error creating category notification:', error);
            throw error;
        }
    },

    // Tạo notification cho zone
    createZoneNotification: async (action, zoneName, storeId, userId) => {
        try {
            const response = await api.post(`${NOTIFICATION_API}/zone`, {
                action,
                zoneName,
                storeId,
                userId
            });
            return response.data;
        } catch (error) {
            console.error('Error creating zone notification:', error);
            throw error;
        }
    }
};

File: backend/src/main/java/com/farmovo/backend/services/NotificationService.java
package com.farmovo.backend.services;

import com.farmovo.backend.dto.CreateNotificationDto;
import com.farmovo.backend.dto.NotificationDto;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

public interface NotificationService {
    
        // Tạo notification mới
    NotificationDto createNotification(CreateNotificationDto dto, Long userId);
    
    // Tạo notification cho tất cả user trong store
    void createNotificationForAllUsersInStore(CreateNotificationDto dto, Long storeId);
    
    // Lấy tất cả notification của user
    Page<NotificationDto> getUserNotifications(Long userId, Long storeId, Pageable pageable);
    
    // Lấy tất cả notification của store (cho Staff xem)
    Page<NotificationDto> getStoreNotifications(Long storeId, Pageable pageable);
    
    // Lấy tất cả notification của tất cả store (cho Admin xem)
    Page<NotificationDto> getAllStoreNotifications(Pageable pageable);
    
    // Lấy số notification chưa đọc của user
    Long getUnreadCount(Long userId, Long storeId);
    
    // Lấy số notification chưa đọc của store (cho Staff xem)
    Long getStoreUnreadCount(Long storeId);
    
    // Lấy số notification chưa đọc của tất cả store (cho Admin xem)
    Long getAllStoreUnreadCount();
    
    // Đánh dấu notification đã đọc
    void markAsRead(Long notificationId, Long userId);
    
    // Đánh dấu tất cả notification đã đọc
    void markAllAsRead(Long userId, Long storeId);
    
    // Đánh dấu tất cả notification của tất cả store đã đọc (cho Admin)
    void markAllNotificationsAsRead();
    
    // Xóa notification
    void deleteNotification(Long notificationId, Long userId);
    
    // Xóa tất cả notification của user
    void deleteAllUserNotifications(Long userId, Long storeId);
    
    // Xóa notification cũ (scheduled task)
    void cleanupOldNotifications();
    
    // Tạo notification cho các thao tác phổ biến
    void createImportTransactionNotification(String action, String transactionName, Long storeId, Long userId);
    void createImportTransactionNotification(String action, String transactionName, Long storeId, Long userId, String newStatus);
    
    void createSaleTransactionNotification(String action, String transactionName, Long storeId, Long userId);
    void createSaleTransactionNotification(String action, String transactionName, Long storeId, Long userId, String newStatus);
    
    void createProductNotification(String action, String productName, Long storeId, Long userId);
    
    void createCustomerNotification(String action, String customerName, Long storeId, Long userId);
    
    void createStocktakeNotification(String action, String stocktakeName, Long storeId, Long userId);
    void createStocktakeNotification(String action, String stocktakeName, Long storeId, Long userId, String newStatus);
    
    // Tạo notification cho Category (Danh mục)
    void createCategoryNotification(String action, String categoryName, Long storeId, Long userId);
    
    // Tạo notification cho Zone (Khu vực)
    void createZoneNotification(String action, String zoneName, Long storeId, Long userId);
}
